Features/3525 edit notes

philipkcl opened the pull request:
# Make notes editable by the user that created them or an admin

*Please don't delete any sections when completing this PR template; instead enter **N/A** for checkboxes or sections which are not applicable, unless otherwise stated below*

See
* https://github.com/DOAJ/doajPM/issues/3525

Describe the scope/purpose of the PR here in as much detail as you like

## Categorisation

This PR...
- [ ] has scripts to run
- [ ] has migrations to run
- [ ] adds new infrastructure
- [ ] changes the CI pipeline
- [ ] affects the public site
- [x] affects the editorial area
- [ ] affects the publisher area
- [ ] affects the monitoring

## Basic PR Checklist

Instructions for developers:
* For each checklist item, if it is N/A to your PR check the N/A box
* For each item that you have done and confirmed for yourself, check Developer box (including if you have checked the N/A box)

Instructions for reviewers:
* For each checklist item that has been confirmed by the Developer, check the Reviewer box if you agree
* For multiple reviewers, feel free to add your own checkbox with your github username next to it if that helps with review tracking

### Code Style

- No deprecated methods are used
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer

- No magic strings/numbers - all strings are in `constants` or `messages` files
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer
  
- ES queries are wrapped in a Query object rather than inlined in the code
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer
  
- Where possible our common library functions have been used (e.g. dates manipulated via `dates`)
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer
  
- Cleaned up commented out code, etc
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer

### Testing

- Unit tests have been added/modified
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer
  
- Functional tests have been added/modified
  - [ ] N/A
  - [ ] Developer
  - [x] Reviewer
  
- Code has been run manually in development, and functional tests followed locally
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer

### Documentation

- FeatureMap annotations have been added
  - [x] N/A
  - [ ] Developer
  - [ ] Reviewer
  
- Documentation updates - if needed - have been identified and prepared for inclusion into main documentation (e.g. added and highlighted/commented as appropriate to this PR)
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer
  
- Core model documentation has been added to if needed: https://docs.google.com/spreadsheets/d/1lun2S9vwGbyfy3WjIjgXBm05D-3wWDZ4bp8xiIYfImM/edit
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer

- Events and consumers documentation has been added if needed: https://docs.google.com/spreadsheets/d/1oIeG5vg-blm2MZCE-7YhwulUlSz6TOUeY8jAftdP9JE/edit
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer
  
- The docs for this branch have been generated and pushed to the doc site (see docs/README.md for details)
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer


### Release Readiness

- If needed, migration has been created and tested locally
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer

- Release sheet has been created, and completed as far as is possible https://docs.google.com/spreadsheets/d/1Bqx23J1MwXzjrmAygbqlU3YHxN1Wf7zkkRv14eTVLZQ/edit#gid=1177753675
  - [ ] N/A
  - [ ] Developer
  - [x] Reviewer

- There has been a recent merge up from `develop` (or other base branch).  List the dates of the merges up from develop below
  - 2023-04-03


## Testing

List the Functional Tests that must be run to confirm this feature:

doajtest/testbook/journal_form/associate_form.yml
doajtest/testbook/journal_form/editor_form.yml
doajtest/testbook/journal_form/maned_form.yml
doajtest/testbook/new_application_form/associate_editor_form.yml
doajtest/testbook/new_application_form/editor_form.yml
doajtest/testbook/new_application_form/maned_form.yml

(just need to wait for the testbook deploy to the docs server to work and then we can replace this with the actual links)


## Deployment

What deployment considerations are there? (delete any sections you don't need)

### Configuration changes

What configuration changes are included in this PR, and do we need to set specific values for production

### Scripts

What scripts need to be run from the PR (e.g. if this is a report generating feature), and when (once, regularly, etc).

### Migrations

What migrations need to be run to deploy this

### Monitoring

What additional monitoring is required of the application as a result of this feature

### New Infrastructure

What new infrastructure does this PR require (e.g. new services that need to run on the back-end).

### Continuous Integration

What CI changes are required for this

richard-jones commented on portality/forms/application_forms.py:
```diff
@@ -1802,6 +1804,20 @@ class FieldDefinitions:
+    }
+
+    # ~~->$ NoteAuthor:FormField~~
+    NOTE_AUTHOR = {
```
Suggest we also add a placeholder text here for legacy notes where there is no author, that says something like "Original author unknown".  Placeholders can be added in `help.placeholder`, you can see an example in something like the `OA_STATEMENT_URL`

richard-jones commented on portality/crosswalks/journal_form.py:
```diff
@@ -430,7 +433,11 @@ def bibjson2form(cls, bibjson, forminfo):
-            note_obj = {'note': n['note'], 'note_date': n['date'], 'note_id': n['id']}
+            author_id = n.get('author_id', '')
+            note_obj = {'note': n['note'], 'note_date': n['date'], 'note_id': n['id'],
+                        'note_author': Account.get_name_safe(author_id) or author_id,
```
I suggest for `note_author` we add both the name and the user account id, so that the editor can find that user by their id if they need to.  For example

```
'note_author': Account.get_name_safe(author_id) + "(" + author_id + ")"
```

philipkcl replied:
fixed with placeholder
![image](https://user-images.githubusercontent.com/97410502/231436857-18eb9715-8766-4c42-9cff-65a823c68a64.png)

richard-jones reviewed (changes requested):
I've reviewed the code and I've played around with the feature.  Overall everything looks great, this is very close to done from the development side, I believe.

I have added a couple of code style/method naming comments which I think would make the code better for the long term (though they do not impact on the functional behaviour of the code).

In reviewing the UI, I have made some suggestions for changes to improve how it looks to the user, and I've added those comments inline in the code comments.  Only one thing didn't fit anywhere:

Could we make it so that when you "Add" a new note, that the user field is populated with something.  This could be:

* The user's actual name/user id
* The text "You are adding a note"

Finally, we should ask @sssoz about the ordering of the fields.  Part of me wants to move the name and the date to *below* the text box, but I'm not sure if that's better.

philipkcl commented on portality/models/account.py:
```diff
@@ -235,3 +236,11 @@ def new_short_uuid(cls):
             return cls.new_short_uuid()
+
+    @classmethod
+    def get_name_safe(cls, account_id) -> str:
```
`_safe` is important, it's not design for jinja template, but it have following features:
* high level API
* defensive / robust design 
* always return string
* will not throw exception if account not found 
* easy to use 
* non flexible

`high level API` which mean it easy to use. for example, if developer just want to get name without handling account not found or return None, than you should use `get_name_safe`
`high level API` which also mean that it's not flexible like `low level APi`, for example, developer will never know what is `''` meaning,  because `''` could be account not found or author.name = ''

in case we need a lower level of `get_name` in the future, I can't rename `get_name_safe` in order to avoid confusing.

richard-jones replied:
ok, seems like a good argument :)

philipkcl commented on portality/view/admin.py:
```diff
@@ -461,7 +461,9 @@ def application_quick_reject(application_id):
     #edit_form = edit_formcontext.render_template()
-    edit_formulaic_context = JournalFormFactory.context("bulk_edit")
+    edit_formulaic_context = JournalFormFactory.context("bulk_edit", extra_param={
+        'cur_user': current_user,
```
I would suggest keep `cur_user` remain unchanged.
because `disable_edit_note_except_cur_user` only accept current user that reading or generating that form.

that can avoid caller (flask, script or background job) wrongly input some other user like account of Journal creator  or account of Journal editor .

richard-jones replied:
ok, in which case let's be specific about the kind of user without using Flask semantics.  Let's call it something like `editing_user`.

amdomanska commented:
I tested the PR as a user without the admin privileges. I added the note to the application assigned to me and saved it but I couldn't edit this note later. I understand it should be possible? As an admit I could edit all the notes.

philipkcl commented:
reproduced and it not working in application review
* author id have not be assigned to note
* and note will disappear after refresh

philipkcl commented:
I can't reproduce that problem

amdomanska commented:
I checked and can see the same problem on develop with some ES exceptions. Means it is most likely local problem for me. Would be useful just for someone else to double-check it @Steven-Eardley @richard-jones

richard-jones commented on portality/lib/formulaic.py:
```diff
@@ -515,6 +534,28 @@ def __init__(self, definition, parent):
         self._definition = definition
         self._formulaic_fieldset = parent
 
+        self.wtfinst : Optional[Field] = None  # used by some plugins function
```
Yes, Formulaic is quite complex, part of me would like to refactor it now that we've had a chance to use it for a while, but I'm not sure there's time.

I'm happy to take your argument above to leave it as-is, and perhaps we can make the whole thing a little less complex in a future iteration :)

richard-jones reviewed (changes requested):
Pending a review from @sssoz on the UI this is basically ready to go to test.

I've marked it "Request changes" because we need to add some functional tests for this.  Propose that we extend the existing application form tests, such as:

* https://doaj.github.io/doaj-docs/develop/testbook/index.html#journal_form/maned_form/journal_form_test
* https://doaj.github.io/doaj-docs/develop/testbook/index.html#journal_form/editor_form/interactions
* https://doaj.github.io/doaj-docs/develop/testbook/index.html#journal_form/associate_form/interactions
* https://doaj.github.io/doaj-docs/develop/testbook/index.html#new_application_form/associate_editor_form
* https://doaj.github.io/doaj-docs/develop/testbook/index.html#new_application_form/editor_form
* https://doaj.github.io/doaj-docs/develop/testbook/index.html#new_application_form/maned_form/form_overview

I'd probably just add a short new test to each of those, which would confirm the following behaviours:

1. If you are an admin you can edit any note, even those that you didn't write yourself
2. If you are not an admin you cannot edit other people's notes
3. If you are not an admin you can only edit your own note

@philipkcl if you are uncomfortable writing the functional tests let me know, and I can add them

philipkcl commented:
@richard-jones 
I have updated PR and added some functional test .
but I'm not sure am i write functional test correctly.
Do I need to run `testbook` cmd ? if yes, can you show me which parameter should I use in `testbook` cmd?

Also @amdomanska , find out that `edit note` not working in `Review application`,  the note can't even able to be saved in application form of her environment, I can't reproduce that problem, do you have the same issues when create note in `Review application`?

richard-jones reviewed (approved):
This looks all good to me.  I slightly extended the functional tests in preparation for DOAJ team testing.

philipkcl commented:
please use / merge following PR that have correct branch name
* https://github.com/DOAJ/doaj/pull/2238
